### ISSUE TITLE
compile straight to jar

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#5445]
There used to be plugin for compiling straight to JAR - https://github.com/typesafehub/sbt-tojar. It might be good to bring this into sbt.

Ref https://github.com/sbt/sbt/issues/2293

**Comment by tmimmanuel** [comment]
@eed3si9n I'd like work on this issue. Please let me if it's available. Thanks.

**Comment by SethTisue** [comment]
@lrytz do I remember right that there's history on this in a customer context...?

**Comment by lrytz** [comment]
Yes, I recently summarized it here: https://github.com/scala/scala3/pull/25215#issuecomment-3996659652

**Comment by eed3si9n** [comment]
If I recall correctly, the current implementation is sort of append-only for performance reason, so if someone deleted a class during the incremental compilation, and then published the JAR, the zip file would technically contain the blob for the deleted class, which is a tradeoff okay for company-internal build tool but may not be for remote-cache-by-default sbt 2.x, as it stands?

**Comment by SethTisue** [comment]
My sense is that it's better to keep it exiled to sbt-tojar where companies that care (if there are any? the company I alluded to above doesn't use sbt) can fund or contribute targeted work on it, rather than add the maintenance burden to the main sbt repo.

**Comment by eed3si9n** [comment]
Yea, set aside the append-only tradeoff, I think that there is a performance benefit of not creating `*.class` files if the eventual destination is `*.jar` anyway. Or at least that was my thinking I guess back in 2020. Let's close this for now, and if ppl want to they can try to bring back the plugin first.

**Comment by lrytz** [comment]
> the current implementation is sort of append-only

Yes, I think so. The jar built by zinc in an incremental compilation contains data for old classfiles, they are just not referenced from the zip index.

The optimus buildtool has a post-processing step to build a new jar after that which only contains referenced data. This is done without de-compressing data. Source [is here](https://github.com/morganstanley/optimus-cirrus/blob/main/optimus/buildtool/projects/app/src/main/scala/optimus/buildtool/utils/TruncatingIndexBasedZipOps.scala) (Apache 2.0 licensed).

They also store hashes for every entry in META-INF, compute those hashes for new entries added in an incremental compilation, and store a global hash in META-INF to quickly tell if two jars contains the same class files.
